### PR TITLE
Experimental: VoidResult

### PR DESCRIPTION
**Pull request by grant-d**
An empty struct, used to represent **void** in generic types.

### CODE REVIEW

File: src/SourceCode.Clay/Discriminated.cs [hunk 1]
```diff
@@ -42,36 +41,6 @@ namespace SourceCode.Clay
         /// </summary>
         public TItem2 Item2 => _state == 2 ? _item2 : throw new InvalidOperationException("The second item is not set.");
 
-        [ExcludeFromCodeCoverage]
```
Comment (grant-d) [inline src/SourceCode.Clay/Discriminated.cs]: Dead code

File: src/SourceCode.Clay/Discriminated.cs [hunk 2]
```diff
@@ -13,7 +12,7 @@ namespace SourceCode.Clay
     [DebuggerDisplay("{Which,nq} {Value}")]
     public readonly struct Discriminated<TItem1, TItem2> : IEquatable<Discriminated<TItem1, TItem2>>
     {
-        private readonly int _state;
+        private readonly byte _state;
```
Comment (grant-d) [inline src/SourceCode.Clay/Discriminated.cs]: Micro optimization

File: src/SourceCode.Clay/VoidResult.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+// Copyright (c) K2 Workflow (SourceCode Technology Holdings Inc.). All rights reserved.
+// Licensed under the MIT License. See LICENSE file in the project root for full license information.
+
+#endregion
+
+using System;
+using System.Runtime.InteropServices;
+
+// Idea derived from various discussions, not limited to:
```
Comment (grant-d) [inline src/SourceCode.Clay/VoidResult.cs]: Motivation

File: src/SourceCode.Clay.Javascript/Ast/JSArrayExpression.cs [hunk 4]
```diff
@@ -8,22 +8,42 @@ public class JSArrayExpression : JSExpression, IEnumerable
     {
         public override JSNodeType Type => JSNodeType.ArrayExpression;
 
-        public List<JSExpression> Elements { get; }
+        public IList<JSExpression> Elements { get; }
```
Comment (jcdickinson) [inline src/SourceCode.Clay.Javascript/Ast/JSArrayExpression.cs]: Revert all of these. IList does not have AddRange.

Comment (jcdickinson) [reply src/SourceCode.Clay.Javascript/Ast/JSArrayExpression.cs]: Not in the code where I am using these types. The interface is insufficient.

**Pull request merged**